Adding v4 Kernel loop explanation (#2400)
OceanPARCELS/parcels: docs/user_guide/examples/explanation_kernelloop.md

---
file_format: mystnb
kernelspec:
  name: python3
---

# The Parcels Kernel loop

On this page we discuss Parcels' execution loop, and what happens under the hood when you combine multiple Kernels.

This is probably not very relevant when you only use the built-in Advection kernels, but can be important when you are writing and combining your own Kernels!

## Background

> When you run a Parcels simulation (i.e. a call to `pset.execute()`), the Kernel loop is the main part of the code that is executed. This part of the code loops through all particles and executes the Kernels that are defined for each particle.

Reviewer: ```suggestion
When you run a Parcels simulation (i.e. a call to `pset.execute()`), the Kernel loop is the main part of the code that is executed. This part of the code loops through time and executes the Kernels for all particle.
```

    out_of_bounds = particles.state == StatusCode.ErrorOutOfBounds
    particles[out_of_bounds].state = StatusCode.Delete

def DeleteAnyError(particles, fieldset):
    any_error = particles.state >= 50  # This captures all Errors
    particles[any_error].state = StatusCode.Delete
```

But of course, you can also write code for more sophisticated behaviour than just deleting the particle. It's up to you! Note that if you don't delete the particle, you will have to update the `particles.state = StatusCode.Success` yourself. For example:

```
def Move1DegreeWest(particles, fieldset):
    out_of_bounds = particles.state == StatusCode.ErrorOutOfBounds
    particles[out_of_bounds].dlon -= 1.0
    particles[out_of_bounds].state = StatusCode.Success
```

Or, if you want to make sure that particles don't escape through the water surface

```
def KeepInOcean(particles, fieldset):
    through_surface = particles.state == StatusCode.ErrorThroughSurface
    particles[through_surface].dz = 0.0
    particles[through_surface].state = StatusCode.Success
```

Kernel functions such as the ones above can then be added to the list of kernels in `pset.execute()`.

> Note that these Kernels that control what to do with `particles.state` should typically be added at the _end_ of the Kernel list, because otherwise later Kernels may overwrite the `particles.state` or the `particles.dlon` variables.

Reviewer: Make this a note (with a box around it)?
